pkg/cache/v2/cache.go

```diff
@@ -37,14 +37,14 @@ type ConfigWatcher interface {
 	// CreateWatch returns a new open watch from a non-empty request.
 	// An individual consumer normally issues a single open watch by each type URL.
 	//
-	// Value channel produces requested resources, once they are available.  If
-	// the channel is closed prior to cancellation of the watch, an unrecoverable
-	// error has occurred in the producer, and the consumer should close the
+	// Once the requested resources are available, they should be pushed into responses channel.
+	// If there are any failures - the error should be returned and the consumer should close the
 	// corresponding stream.
+	// Cache implementation should not close responses channel.
 	//
 	// Cancel is an optional function to release resources in the producer. If
 	// provided, the consumer may call this function multiple times.
-	CreateWatch(*Request) (value chan Response, cancel func())
+	CreateWatch(request *Request, responses chan<- Response) (cancel func(), err error)
```

Comment:
This isn't backwards compatible and will break a bunch of server implementations - possible to implement as a separate API? @kyessenov May be worth discussing major/minor versions for go-control-plane as adoption increases.

Reply:
For some more context, we have a management server implementation that relies on closing the channel in order to terminate the ack/nack flow from the client, forcing client-issued retries. Under this model, is that capability lost?

Reply:
@jessicayuen There is a bunch of discussion why the change seems like a good idea [here](https://github.com/envoyproxy/go-control-plane/pull/353). Stable versions are important, but we kept it at 0.x.y because we felt it's better to give time to iron out issues in APIs. We should focus on getting 1.0 at some point.

There is an error returned in the interface stub. I think the difference is deferred signalling of errors. Does you cache implementation close the channel asyncrhonously?

Reply:
In the xds-relay implementation, channels are closed asynchronously. My main concern with this approach is that the ownership is moving. The implementor no longer has control over the channel being open and runs the risk of writing to a closed channel. It's also not apparent what the implementor should do in the event the channel is closed - to return an error? I'll follow up more with the discussions over the weekend.